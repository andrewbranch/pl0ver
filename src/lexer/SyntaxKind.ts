export enum SyntaxKind {
  BadToken,
  CloseParenthesisToken,
  ColonEqualToken,
  CommaToken,
  DotToken,
  EOFToken,
  EqualToken,
  GreaterThanToken,
  GreaterThanEqualToken,
  HashToken,
  LessThanToken,
  LessThanEqualToken,
  MinusToken,
  OpenParenthesisToken,
  PlusToken,
  SemicolonToken,
  SlashToken,
  StarToken,
  QuestionToken,
  
  BeginKeyword,
  CallKeyword,
  ConstKeyword,
  DoKeyword,
  EndKeyword,
  IfKeyword,
  OddKeyword,
  ProcedureKeyword,
  ThenKeyword,
  WhileKeyword,
  VarKeyword,

  NumericLiteral,

  Identifier,

  NewlineTrivia,
  WhitespaceTrivia,
}