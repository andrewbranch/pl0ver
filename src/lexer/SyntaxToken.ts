import { SyntaxKind } from './SyntaxKind';

export class SyntaxToken {
  private __debugKind: string | undefined;
  constructor(public kind: SyntaxKind, public value?: string) {
    if (process.env.NODE_ENV !== 'production') {
      this.__debugKind = SyntaxKind[kind];
    }
  }
}