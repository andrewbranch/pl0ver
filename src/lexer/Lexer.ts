import { SyntaxKind } from './SyntaxKind';
import { SyntaxToken } from './SyntaxToken';
import { isSingleLineWhitespaceCharacter, isNewlineCharacter, isNumeral, isLetter, isIdentifierCharacter } from '../util/text';

export class Lexer {
  private position = 0;
  public text = '';

  constructor(text: string) {
    this.setText(text);
  }

  public lex(): SyntaxToken {
    if (this.position >= this.text.length) {
      return new SyntaxToken(SyntaxKind.EOFToken);
    }

    let text = this.advance();
    let kind = SyntaxKind.BadToken;

    // Tokens
    if (text === '(') {
      kind = SyntaxKind.OpenParenthesisToken;
    } else if (text === ')') {
      kind = SyntaxKind.CloseParenthesisToken;
    } else if (text === ',') {
      kind = SyntaxKind.CommaToken;
    } else if (text === '.') {
      kind = SyntaxKind.DotToken;
    } else if (text === '>') {
      if (this.getText() === '=') {
        text += this.advance();
        kind = SyntaxKind.GreaterThanEqualToken;
      } else {
        kind = SyntaxKind.GreaterThanToken;
      }
    } else if (text === '<') {
      if (this.getText() === '=') {
        text += this.advance();
        kind = SyntaxKind.LessThanEqualToken;
      } else {
        kind = SyntaxKind.LessThanToken;
      }
    } else if (text === '#') {
      kind = SyntaxKind.HashToken;
    } else if (text === ':') {
      if (this.getText() === '=') {
        text += this.advance();
        kind = SyntaxKind.ColonEqualToken;
      }
    } else if (text === '=') {
      kind = SyntaxKind.EqualToken;
    } else if (text === '-') {
      kind = SyntaxKind.MinusToken;
    } else if (text === '+') {
      kind = SyntaxKind.PlusToken;
    } else if (text === '*') {
      kind = SyntaxKind.StarToken;
    } else if (text === '/') {
      kind = SyntaxKind.SlashToken;
    } else if (text === ';') {
      kind = SyntaxKind.SemicolonToken;
    } else if (text === '?') {
      kind = SyntaxKind.QuestionToken;
    }

    // Trivia
    else if (isSingleLineWhitespaceCharacter(text)) {
      kind = SyntaxKind.WhitespaceTrivia;
      text += this.lexWhitespaceTrivia();
    } else if (isNewlineCharacter(text)) {
      kind = SyntaxKind.NewlineTrivia;
      text += this.lexNewlineTrivia();
    }

    // Literals
    else if (isNumeral(text)) {
      kind = SyntaxKind.NumericLiteral;
      text += this.lexNumericLiteral();
    }

    // Keywords and identifiers
    else if (isLetter(text)) {
      text += this.lexIdentifier();
      switch (text.toLowerCase()) {
        case 'begin':
          kind = SyntaxKind.BeginKeyword;
          break;
        case 'call':
          kind = SyntaxKind.CallKeyword;
          break;
        case 'const':
          kind = SyntaxKind.ConstKeyword;
          break;
        case 'do':
          kind = SyntaxKind.DoKeyword;
          break;
        case 'end':
          kind = SyntaxKind.EndKeyword;
          break;
        case 'if':
          kind = SyntaxKind.IfKeyword;
          break;
        case 'odd':
          kind = SyntaxKind.OddKeyword;
          break;
        case 'procedure':
          kind = SyntaxKind.ProcedureKeyword;
          break;
        case 'then':
          kind = SyntaxKind.ThenKeyword;
          break;
        case 'while':
          kind = SyntaxKind.WhileKeyword;
          break;
        case 'var':
          kind = SyntaxKind.VarKeyword;
          break;
        default:
          kind = SyntaxKind.Identifier;
      }
    }

    return new SyntaxToken(kind, text);
  }

  private setText(text: string) {
    this.text = text;
  }

  private getText(length = 1) {
    return this.text.slice(this.position, this.position + length);
  }

  private advanceWhile(predicate: (text: string) => boolean): string {
    let text = '';
    while (predicate(this.getText())) {
      text += this.advance();
    }
    return text;
  }

  private advance(by = 1): string {
    const text = this.getText(by);
    this.position += by;
    if (this.position > this.text.length) {
      throw new Error('Cannot advance past end of input.');
    }
    return text;
  }

  private lexWhitespaceTrivia(): string {
    return this.advanceWhile(char => isSingleLineWhitespaceCharacter(char));
  }

  private lexNewlineTrivia(): string {
    return this.advanceWhile(char => isNewlineCharacter(char));
  }

  private lexNumericLiteral(): string {
    let text = this.advanceWhile(char => isNumeral(char));
    if (this.getText() === '.') {
      text += this.advance();
      text += this.advanceWhile(char => isNumeral(char));
    }
    return text;
  }

  private lexIdentifier(): string {
    return this.advanceWhile(char => isIdentifierCharacter(char));
  }
}