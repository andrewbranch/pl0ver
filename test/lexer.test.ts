import fs from 'fs';
import path from 'path';
import { Lexer } from '../src/lexer/Lexer';
import { SyntaxToken } from '../src/lexer/SyntaxToken';
import { SyntaxKind } from '../src/lexer/SyntaxKind';

const sample1 = fs.readFileSync(path.join(__dirname, 'fixtures/sample-1.txt'), 'utf8');

describe('Lexer', () => {
  it('works', () => {
    const lexer = new Lexer(sample1);
    const syntaxList: SyntaxToken[] = [];
    while (true) {
      const token = lexer.lex();
      syntaxList.push(token);
      if (token.kind === SyntaxKind.EOFToken) {
        break;
      }
    }

    expect(syntaxList).toMatchSnapshot();
  });
});
