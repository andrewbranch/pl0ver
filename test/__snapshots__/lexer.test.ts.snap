// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lexer works 1`] = `
Array [
  SyntaxToken {
    "__debugKind": "VarKeyword",
    "kind": 29,
    "value": "VAR",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "x",
  },
  SyntaxToken {
    "__debugKind": "CommaToken",
    "kind": 3,
    "value": ",",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "squ",
  },
  SyntaxToken {
    "__debugKind": "SemicolonToken",
    "kind": 15,
    "value": ";",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "

",
  },
  SyntaxToken {
    "__debugKind": "ProcedureKeyword",
    "kind": 26,
    "value": "PROCEDURE",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "square",
  },
  SyntaxToken {
    "__debugKind": "SemicolonToken",
    "kind": 15,
    "value": ";",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "BeginKeyword",
    "kind": 19,
    "value": "BEGIN",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": "   ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "squ",
  },
  SyntaxToken {
    "__debugKind": "ColonEqualToken",
    "kind": 2,
    "value": ":=",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "x",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "StarToken",
    "kind": 17,
    "value": "*",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "x",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "EndKeyword",
    "kind": 23,
    "value": "END",
  },
  SyntaxToken {
    "__debugKind": "SemicolonToken",
    "kind": 15,
    "value": ";",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "

",
  },
  SyntaxToken {
    "__debugKind": "BeginKeyword",
    "kind": 19,
    "value": "BEGIN",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": "   ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "x",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "ColonEqualToken",
    "kind": 2,
    "value": ":=",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "NumericLiteral",
    "kind": 30,
    "value": "1",
  },
  SyntaxToken {
    "__debugKind": "SemicolonToken",
    "kind": 15,
    "value": ";",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": "   ",
  },
  SyntaxToken {
    "__debugKind": "WhileKeyword",
    "kind": 28,
    "value": "WHILE",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "x",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "LessThanEqualToken",
    "kind": 11,
    "value": "<=",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "NumericLiteral",
    "kind": 30,
    "value": "10",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "DoKeyword",
    "kind": 22,
    "value": "DO",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": "   ",
  },
  SyntaxToken {
    "__debugKind": "BeginKeyword",
    "kind": 19,
    "value": "BEGIN",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": "      ",
  },
  SyntaxToken {
    "__debugKind": "CallKeyword",
    "kind": 20,
    "value": "CALL",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "square",
  },
  SyntaxToken {
    "__debugKind": "SemicolonToken",
    "kind": 15,
    "value": ";",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": "      ",
  },
  SyntaxToken {
    "__debugKind": "BadToken",
    "kind": 0,
    "value": "!",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "squ",
  },
  SyntaxToken {
    "__debugKind": "SemicolonToken",
    "kind": 15,
    "value": ";",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": "      ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "x",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "ColonEqualToken",
    "kind": 2,
    "value": ":=",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "Identifier",
    "kind": 31,
    "value": "x",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "PlusToken",
    "kind": 14,
    "value": "+",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": " ",
  },
  SyntaxToken {
    "__debugKind": "NumericLiteral",
    "kind": 30,
    "value": "1",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "WhitespaceTrivia",
    "kind": 33,
    "value": "   ",
  },
  SyntaxToken {
    "__debugKind": "EndKeyword",
    "kind": 23,
    "value": "END",
  },
  SyntaxToken {
    "__debugKind": "NewlineTrivia",
    "kind": 32,
    "value": "
",
  },
  SyntaxToken {
    "__debugKind": "EndKeyword",
    "kind": 23,
    "value": "END",
  },
  SyntaxToken {
    "__debugKind": "DotToken",
    "kind": 4,
    "value": ".",
  },
  SyntaxToken {
    "__debugKind": "EOFToken",
    "kind": 5,
    "value": undefined,
  },
]
`;
